pkg/document/document.go

```diff
@@ -1,60 +1,298 @@
+
+func NewDocument(u uri.URI, input []byte, tree *sitter.Tree) Document {
+	return &document{
+		uri:   u,
+		input: input,
+		tree:  tree,
+	}
+}
 
-func NewDocument(input []byte, tree *sitter.Tree) Document {
-	return document{
+func NewDocumentWithSymbols(u uri.URI, input []byte, tree *sitter.Tree) Document {
+	doc := &document{
+		uri:   u,
 		input: input,
 		tree:  tree,
 	}
+	doc.functions = query.Functions(doc, tree.RootNode())
+	doc.symbols = query.DocumentSymbols(doc)
+	doc.parseLoadStatements()
+	return doc
+}
+
+func NodesToContent(doc Document, nodes []*sitter.Node) string {
+	var content string
+	if len(nodes) > 0 {
+		content = doc.ContentRange(sitter.Range{
+			StartByte: nodes[0].StartByte(),
+			EndByte:   nodes[len(nodes)-1].EndByte(),
+		})
+	} else {
+		content = doc.Content(doc.Tree().RootNode())
+	}
+	return content
 }
 
 type document struct {
+	uri uri.URI
+
 	// input is the file as it exists in the editor buffer.
 	input []byte
 
 	// tree represents the parsed version of the document.
 	tree *sitter.Tree
+
+	functions   map[string]protocol.SignatureInformation
+	symbols     []protocol.DocumentSymbol
+	diagnostics []protocol.Diagnostic
+	loads       []LoadStatement
 }
 
-var _ Document = document{}
+var _ Document = &document{}
 
-func (d document) Content(n *sitter.Node) string {
+func (d *document) Content(n *sitter.Node) string {
 	return n.Content(d.input)
 }
 
-func (d document) ContentRange(r sitter.Range) string {
+func (d *document) ContentRange(r sitter.Range) string {
 	return string(d.input[r.StartByte:r.EndByte])
 }
 
-func (d document) Tree() *sitter.Tree {
+func (d *document) Tree() *sitter.Tree {
 	return d.tree
 }
 
-func (d document) Close() {
+func (d *document) Functions() map[string]protocol.SignatureInformation {
+	return d.functions
+}
+
+func (d *document) Symbols() []protocol.DocumentSymbol {
+	return d.symbols
+}
+
+func (d *document) Diagnostics() []protocol.Diagnostic {
+	return d.diagnostics
+}
+
+func (d *document) Loads() []LoadStatement {
+	return d.loads
+}
+
+func (d *document) Close() {
 	d.tree.Close()
 }
 
 // Copy creates a shallow copy of the Document.
 //
 // The Contents byte slice is returned as-is.
 // A shallow copy of the Tree is made, as Tree-sitter trees are not thread-safe.
-func (d document) Copy() Document {
-	return document{
-		input: d.input,
-		tree:  d.tree.Copy(),
+func (d *document) Copy() Document {
+	doc := &document{
+		uri:         d.uri,
+		input:       d.input,
+		tree:        d.tree.Copy(),
+		functions:   make(map[string]protocol.SignatureInformation),
+		symbols:     append([]protocol.DocumentSymbol{}, d.symbols...),
+		loads:       append([]LoadStatement{}, d.loads...),
+		diagnostics: append([]protocol.Diagnostic{}, d.diagnostics...),
+	}
+	for fn := range d.functions {
+		doc.functions[fn] = d.functions[fn]
+	}
+	return doc
+}
+
+func (d *document) processLoads(ctx context.Context, m *Manager) {
+	for i, load := range d.loads {
+		if load.File == "" {
```

Comment:
we should probably also skip files that start with "ext://" (or maybe contain "://" at all)

(fwiw, if referring to "ext://" as a "file" feels weird - the starlark spec resolves this by referring to the first arg as the "module")